BlueprintUI/Sources/Environment/Environment.swift

```diff
@@ -34,33 +34,103 @@ import UIKit
 ///             }
 ///         }
 ///     }
-public struct Environment {
+public struct Environment: Equatable {
+
+    typealias OnDidRead = (StorageKey) -> Void
+
+    enum LayoutPass: Equatable {
+        case measurement
+        case layout
+    }
 
     /// A default "empty" environment, with no values overridden.
     /// Each key will return its default value.
     public static let empty = Environment()
 
-    private var values: [ObjectIdentifier: Any] = [:]
+    /// When enabled, notify any subscribers when a value is read
+    /// from the environment
+    var readNotificationsEnabled: Bool = true
+
+    /// If the `Environment` contains any values.
+    var isEmpty: Bool {
+        values.isEmpty
+    }
+
+    private var values: [StorageKey: ValueBox] = [:]
+
+    private var measurementDidRead: OnDidRead?
+    private var layoutDidRead: [OnDidRead] = []
```

Comment:
A very good question!

We differentiate these because measurements are "flat", eg we only care about the dependencies read during the direct element's measurement – because every layer of the tree that can caches its own measurements. However for layouts, we cache entire trees of layout – including our children – so we want to record and track the environment dependencies of our children as well, and re-perform the layouts if those values change as well.

Does that make sense? I feel like I should probably describe this both better in the code docs here, and in some explanatory tests

Reply:
> Does that make sense?

I think so!

> I feel like I should probably describe this both better in the code docs here, and in some explanatory tests

+ 1

---

If we used an array to describe the measurement subscriptions (just to simplify the implementation a bit within environment and perhaps prepare the shape better for supporting "scopes" rather than `enum LayoutPass`) would we need to manually clear them out for some reason, or would there just end up being only 1 subscription of that scope?

Reply:
I think we can simplify this to a single array!

Reply:
Thinking forward through this more, we might actually want to simplify measurement to be append as well. In theory, once we allow measurements and layouts to read the environment, downstream elements could pull a value from the environment, which when the parent measures its children, would affect their size.